### PR TITLE
feat: add bee desktop toolkit

### PR DESCRIPTION
**Pull request by Cafe137**

### CODE REVIEW

**Comment by bee-worker** [comment]
## 🐝 PR preview in Swarm

**Preview URL:** https://bah5acgzafkpycdkf5ju7znbkxnaquecoaz6ch4ik65zzmfqqvjl37l5elrda.bzz.link
**Swarm Hash:** `2a9f810d45ea69fcb42abb410a104e067c23f10af773961610aa57bfafa45c46`
**Commit Hash:** c1a5ec555dd5244711fba76c5edb5e28c3bf4f7a
**Commit Message:** feat: add light node upgrade

**Comment by vojtechsimetka** [comment]
@Cafe137 what is your idea about this? Should we be rebasing this on master continuously (maybe after we shave off some of the already implemented functionality)? Or do you want to merge this?

File: src/layout/Dashboard.tsx [hunk 1]
```diff
@@ -33,7 +29,6 @@ const Dashboard = (props: Props): ReactElement => {
       <Container className={classes.content}>
         <ErrorBoundary>
           <>
-            <AlertVersion />
```
Comment (Cafe137) [inline src/layout/Dashboard.tsx:36]: What should we do about AlertVersion?

File: src/utils/rpc.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { Contract, providers } from 'ethers'
+
+const PROVIDER = 'https://gno.getblock.io/mainnet/?api_key=d7b92d96-9784-49a8-a800-b3edd1647fc7'
```
Comment (Cafe137) [inline src/utils/rpc.ts:3]: I'm OK with this until we have FDS RPC working from browser

File: src/layout/Dashboard.tsx (hunk 1 above)
Comment (vojtechsimetka) [reply src/layout/Dashboard.tsx:36]: IMHO it should stay in general but we should make it less strict.

File: src/providers/Bee.tsx [hunk 3]
```diff
@@ -61,6 +61,14 @@ interface ContextInterface {
...
 }
 
+const urlSearchParams = new URLSearchParams(window.location.search)
+const newApiKey = urlSearchParams.get('v')
+
+if (newApiKey) {
+  localStorage.setItem('apiKey', newApiKey)
+  window.location.search = ''
+}
+
```
Comment (vojtechsimetka) [inline src/providers/Bee.tsx:71]: Could you extract this into `SettingsProvider` and store there? That is the top level provider and IMHO makes sense to also store the key in info. It already passes down the `Bee` object.

File: src/utils/desktop.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+export async function getBeeEthereumAddress(): Promise<string> {
+  const status = await getDesktopStatus()
+
+  if (!status.address) {
+    throw Error('Bee node Ethereum address is not yet available')
+  }
+
+  return status.address
+}
```
Comment (vojtechsimetka) [inline src/utils/desktop.ts:24]: This is a duplication, the Bee provider already has the ethereum address. I am fine with it, just wanted to mention this.

File: src/utils/net.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  const authorization = localStorage.getItem('apiKey')
+
+  if (!authorization) {
+    throw Error('API key not found in local storage')
+  }
+  const headers = {
+    'content-type': 'application/json',
+    authorization,
+  }
+  const response = await fetch(url, {
```
Comment (vojtechsimetka) [inline src/utils/net.ts:25]: can we just use `axios`? We use it elsewhere.

File: src/utils/rpc.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+async function eth_getBalance(address: string): Promise<string> {
+  if (!address.startsWith('0x')) {
+    address = `0x${address}`
+  }
+  const response = await fetch(PROVIDER, {
```
Comment (vojtechsimetka) [inline src/utils/rpc.ts:9]: can you use axios?

File: src/providers/Bee.tsx [hunk 7]
```diff
@@ -209,6 +222,22 @@ export function Provider({ children }: Props): ReactElement {
...
+      if (xdai?.then) {
+        xdai.then(balance => setXdai(balance))
+      }
+
+      if (bzz?.then) {
+        bzz.then(balance => setBzz(balance))
+      }
+    }
+  }, [nodeAddresses])
+
```
Comment (Cafe137) [inline src/providers/Bee.tsx:240]: @vojtechsimetka It seemed a good idea to put the RPC balance checks in a `useEffect` hook depending on `nodeAddresses`. However that triggered many duplicate requests at the same time, so I had to introduce debouncing. Lmk what you think of this approach.

File: src/utils/desktop.ts (hunk 4 above)
Comment (Cafe137) [reply src/utils/desktop.ts]: Removed it, no longer needed

File: src/providers/Bee.tsx (hunk 7 above)
Comment (vojtechsimetka) [reply src/providers/Bee.tsx:240]: Yeah this is good. But I think even putting `[nodeAddresses?.ethereum]` in the hook could solve it (so at the beginning it is null and then always the same address. But I am also ok with what you implemented.

File: src/providers/Bee.tsx [hunk 8]
```diff
@@ -7,16 +8,21 @@ import {
...
 import { useLatestBeeRelease } from '../hooks/apiHooks'
 import { Token } from '../models/Token'
 import type { Balance, ChequebookBalance, Settlements } from '../types'
+import { Rpc } from '../utils/rpc'
 import { Context as SettingsContext } from './Settings'
 
+interface RpcBalance {
+  bzz: string
+  xdai: string
+}
```
Comment (vojtechsimetka) [inline src/providers/Bee.tsx:24]: Can these be instances of Token already?

File: src/utils/desktop.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+export async function getDesktopStatus(): Promise<DesktopStatus> {
+  const response = await getJson(`http://${getDesktopHost()}/status`)
+
+  return response as DesktopStatus
+}
+
+export async function getGasFromFaucet(address: string): Promise<void> {
+  await fetch(`http://getxdai.co/${address}/0.1`, {
```
Comment (vojtechsimetka) [inline src/utils/desktop.ts:17]: this still uses fetch

**Review by vojtechsimetka — APPROVED** [review]
Thanks, this looks very good :)

**Pull request merged**